### ISSUE TITLE
when use predict batched function, sometimes break.

### ISSUE THREAD

**Issue by xinsuinizhuan** [josephjaspers/blackcat_tensors#46]
when i test the predict function, some times it beak:
auto batch = inputs[0];
mat hyps = lstmpredicttask->m_pnetwork.predict(batch);
break at:
![图片](https://user-images.githubusercontent.com/40679769/67256613-a9b7cc00-f4ba-11e9-80ba-a6aa2605deca.png)
Call stack info is:
![图片](https://user-images.githubusercontent.com/40679769/67256628-bf2cf600-f4ba-11e9-9632-3fa5cf6d5ef6.png)

**Comment by xinsuinizhuan** [comment]
when i trained, i test this code:
{
		auto batch = inputs[0];
		mat hyps = lstmpredicttask->m_pnetwork.forward_propagation(batch);
		std::cout << " forward_propagation MAPE loss: " << BC::Scalar<double>(BC::nn::MAPE(hyps, outputs[0]) / lstmpredicttask->m_batch_size).data()[0] << std::endl;

		std::cout << "forward_propagation inputdata------------------------------------" << std::endl;
		auto imagesinput0 = reshape(inputs[0], BC::shape(96, 10, lstmpredicttask->m_batch_size));
		imagesinput0[0].t().print_sparse();

		std::cout << "forward_propagation output predict data------------------------------------" << std::endl;
		hyps[0].print();
	}

	{
		auto batch = inputs[0];
		mat hyps = lstmpredicttask->m_pnetwork.predict(batch);
		std::cout << " predict MAPE loss: " << BC::Scalar<double>(BC::nn::MAPE(hyps, outputs[0]) / lstmpredicttask->m_batch_size).data()[0] << std::endl;

		std::cout << "predict inputdata------------------------------------" << std::endl;
		auto imagesinput0 = reshape(inputs[0], BC::shape(96, 10, lstmpredicttask->m_batch_size));
		imagesinput0[0].t().print_sparse();

		std::cout << "predict output predict data------------------------------------" << std::endl;
		hyps[0].print();
	}
break at this line:
mat hyps = lstmpredicttask->m_pnetwork.predict(batch);

**Comment by josephjaspers** [comment]
Just seeing this now I will debug it soon.

**Comment by josephjaspers** [comment]
Could you send me more of what you are working on? I can't seem to replicate the issue.

**Comment by xinsuinizhuan** [comment]
> 
> 
> Could you send me more of what you are working on? I can't seem to replicate the issue.

//start train
	LstmPredictTask* lstmpredicttask = new LstmPredictTask();
	if (lstmpredicttask == NULL) {
		return -2;
	}

	//LstmPredictTask lstmpredicttask;
	std::cout << "Neural Network architecture: \n" << lstmpredicttask->m_pnetwork.get_string_architecture() << std::endl;
	lstmpredicttask->m_pnetwork.set_learning_rate(lstmpredicttask->m_learning_rate);
	lstmpredicttask->m_pnetwork.set_batch_size(lstmpredicttask->m_batch_size);

	int training_sets;
	std::pair<cube, cube> data = load_train_data(system_tag, datafilepath, lstmpredicttask, &training_sets);
	cube& inputs = data.first;
	cube& outputs = data.second;

	std::cout <<" training..." << std::endl;
	auto start = std::chrono::system_clock::now();

	std::cout << "imagesinput real data:------------------------------------" << std::endl;
	auto imagesinput = reshape(inputs[0], BC::shape(96, 10, lstmpredicttask->m_batch_size));
	imagesinput[0].t().print_sparse();

	std::cout << "imagesoutput real data:------------------------------------" << std::endl;
	auto imagesoutput = reshape(outputs[0], BC::shape(96, 2, lstmpredicttask->m_batch_size));
	imagesoutput[0].t().print_sparse();

	for (int i = 0; i < epochs; ++i) {
		std::cout << " current epoch: " << i << std::endl;
		for (int j = 0; j < training_sets; ++j) {
			lstmpredicttask->m_pnetwork.forward_propagation(inputs[j]);
			lstmpredicttask->m_pnetwork.back_propagation(outputs[j]);
			lstmpredicttask->m_pnetwork.update_weights();
		}
	}

	//if (strlen(_trainparamsavefile) != 0)
	//{
	//	lstmpredicttask->m_pnetwork.save(_trainparamsavefile); //Uncomment to add saving/loading
	//}

	auto end = std::chrono::system_clock::now();
	clock total = clock(end - start);
	std::cout << " training time: " << total.count() << std::endl;

	{
		auto batch = inputs[0];
		mat hyps = lstmpredicttask->m_pnetwork.forward_propagation(batch);
		std::cout << " forward_propagation MAPE loss: " << BC::Scalar<double>(BC::nn::MAPE(hyps, outputs[0]) / lstmpredicttask->m_batch_size).data()[0] << std::endl;

		std::cout << "forward_propagation inputdata------------------------------------" << std::endl;
		auto imagesinput0 = reshape(inputs[0], BC::shape(96, 10, lstmpredicttask->m_batch_size));
		imagesinput0[0].t().print_sparse();

		std::cout << "forward_propagation output predict data------------------------------------" << std::endl;
		hyps[0].print();
	}

	{
		auto batch = inputs[0];
		mat hyps = lstmpredicttask->m_pnetwork.predict(batch);
		std::cout << " predict MAPE loss: " << BC::Scalar<double>(BC::nn::MAPE(hyps, outputs[0]) / lstmpredicttask->m_batch_size).data()[0] << std::endl;

		std::cout << "predict inputdata------------------------------------" << std::endl;
		auto imagesinput0 = reshape(inputs[0], BC::shape(96, 10, lstmpredicttask->m_batch_size));
		imagesinput0[0].t().print_sparse();

		std::cout << "predict output predict data------------------------------------" << std::endl;
		hyps[0].print();
	}

**Comment by josephjaspers** [comment]
Is this still occurring? 
I can't seem to replicate it

**Comment by xinsuinizhuan** [comment]
Let me hava a try, right now.

**Comment by xinsuinizhuan** [comment]
I test six times, no occurring. So it seems ok.

**Comment by josephjaspers** [comment]
Awesome!
